### PR TITLE
Fix partial/full filter issue in `service tasks --filter`

### LINKED ISSUE #24108: `docker node task --filter name=` filters on `service`
Taken from the CLI reference page of `docker node tasks`:

> The `name` filter matches on all or part of a task's name.

This is wrong. The filtering is applied on the **exact service name**, see example below.

I think this is actually more useful than filtering on task names and therefore propose to **change the filter name from `name` to `service`**.

Furthermore, filtering should be extended to also cover substring matches, not only exact matches.

```bash
$ docker service create --name top busybox top
7ieif28g1smw7ppav7wiausmz

# a task with name top.1 was created on this node
$ docker node tasks self
ID                         NAME   SERVICE  IMAGE    LAST STATE          DESIRED STATE  NODE
brscnd76m5eh8l3cvmpt5lkj4  top.1  top      busybox  Running 51 seconds  Running        bd4242800842

# filtering by part of the task's name does not find the task
$ docker node tasks --filter name=to self
ID  NAME  SERVICE  IMAGE  LAST STATE  DESIRED STATE  NODE

# filtering by the task's exact name does not find the task
$ docker node tasks --filter name=top.1 self
ID  NAME  SERVICE  IMAGE  LAST STATE  DESIRED STATE  NODE

# ==> only filtering by the SERVICE's name finds the task <==
$ docker node tasks --filter name=top self
ID                         NAME   SERVICE  IMAGE    LAST STATE              DESIRED STATE  NODE
brscnd76m5eh8l3cvmpt5lkj4  top.1  top      busybox  Running About a minute  Running        bd4242800842

$ docker version
Client:
 Version:      1.12.0-dev
 API version:  1.25
 Go version:   go1.6.2
 Git commit:   83e6197-unsupported
 Built:        Tue Jun 28 20:00:19 2016
 OS/Arch:      linux/amd64

Server:
 Version:      1.12.0-dev
 API version:  1.25
 Go version:   go1.6.2
 Git commit:   83e6197-unsupported
 Built:        Tue Jun 28 20:00:19 2016
 OS/Arch:      linux/amd64
```

### LINKED ISSUE #24790: 1.12.0-rc4: inspect task by name (not id)
It would be nice if we can get a task by name as well as id. e.g.:

```
$ docker inspect worker.1
[
    {
        "ID": "bzsrvcl9sr1c74djkf8jl6jvr",
        "Version": {
            "Index": 123
        },
...
]
```

### PR DESCRIPTION
**Pull request by yongtang**
**\- What I did**
This fix tries to address the issue related to #24108 and #24790, and also the case from:
https://github.com/docker/docker/pull/24620#issuecomment-233715656

The reason for the failure case in the above mentioned issues is that currently Task names are actually indexed by Service Name (`e.ServiceAnnotations.Name`), not Task name (`<e.ServiceAnnotations.Name>.<Slot>.<TaskID>`)

**\- How I did it**

To fix it, a pull request in swarmkit  has been opened separately:
https://github.com/docker/swarmkit/pull/1193

Another pull request in engine-api has been opened as well:
https://github.com/docker/engine-api/pull/350

**NOTE**: Based on the discussion in (https://github.com/docker/swarmkit/pull/1193), with the task name defined as `<e.ServiceAnnotations.Name>.<Slot>.<TaskID>`, the output of listing tasks would be (See `NAME` field for differences):

```
ubuntu@ubuntu:~$ sudo docker service ps --filter name=top.1 top
ID                         NAME                                 IMAGE    NODE    DESIRED STATE  CURRENT STATE                ERROR
9k0jjra8lqipvk7k651hkr9h1  top.1.9k0jjra8lqipvk7k651hkr9h1      busybox  ubuntu  Running        Running about a minute ago   
3ue05jmafu3o03lxdjo4k2fd6   \_ top.1.3ue05jmafu3o03lxdjo4k2fd6  busybox  ubuntu  Shutdown       Complete about a minute ago  
a95stag9gel17qhfk81wcv4nq   \_ top.1.a95stag9gel17qhfk81wcv4nq  busybox  ubuntu  Shutdown       Complete 19 minutes ago   
```

Below is the item list to be done based on the discussion in thread (https://github.com/docker/swarmkit/pull/1193).
- [X] Need to assign `<ServiceName>.<Slot>.<TaskID>` to TaskName in orchestrator.
- [X] In  `manager/controlapi/task.go` and `manager/state/store/tasks.go`, use `TaskName` but fallback to calculated`<ServiceName>.<Slot>.<TaskID>`. (**Verified that fallback is required**).
- [X] Update docker's UI/UX so that `NAME` field in listed tasks shows `<ServiceName>.<Slot>.<TaskID>`.
- [X] Test 1.12.0 client against 1.13.0 daemon.
- [X] Test daemon update from 1.12.0 to 1.13.0.

**\- How to verify it**

This fix adds the integration tests for the above mentioned issues. 

**\- Description for the changelog**
**\- A picture of a cute animal (not mandatory but encouraged)**

This fix fixes #24108.
This fix fixes #24790.
This fix is related to #24620.

**NOTE: Swarmkit and engine-api revendoring is needed to completely fix the issues.**

### CODE REVIEW

**Comment by thaJeztah** [comment]
@yongtang looks like this needs a rebase

**Comment by yongtang** [comment]
Thanks @thaJeztah the pull request has been rebased.

**Comment by thaJeztah** [comment]
Looking good; trying this;

Old behavior (filters on service name)

```
docker service create --name web --replicas=5 nginx:alpine
docker service create --name database --replicas=5 nginx:alpine

 docker node ps self
ID                         NAME        IMAGE         NODE  DESIRED STATE  CURRENT STATE          ERROR
6ldq07geg2gsw6w1n6fcngtfo  database.1  nginx:alpine  moby  Running        Running 4 minutes ago
3m0pnnmf2zgu4pmrkxjg0thyl  web.1       nginx:alpine  moby  Running        Running 4 hours ago
e17j7e2trmce62obynvl005zo  database.2  nginx:alpine  moby  Running        Running 4 minutes ago
c4mo4u2dmke38qr14kqhozbgb  web.2       nginx:alpine  moby  Running        Running 5 minutes ago
cy54ecjg7atwt3odyzhsykz3z  database.3  nginx:alpine  moby  Running        Running 4 minutes ago
3idithymwvobg9o3gcqtpj8ci  web.3       nginx:alpine  moby  Running        Running 5 minutes ago
8fjt9pmup6y19xydq2ozyhbwd  database.4  nginx:alpine  moby  Running        Running 4 minutes ago
42l76uqnb0cu5v4ot3q2oeo0j  web.4       nginx:alpine  moby  Running        Running 5 minutes ago
6z19ahl2fox5r3mzhptsit7dc  database.5  nginx:alpine  moby  Running        Running 4 minutes ago
3pwy80otxohhih79s781otbar  web.5       nginx:alpine  moby  Running        Running 5 minutes ago

docker node ps --filter name=web. self
ID  NAME  IMAGE  NODE  DESIRED STATE  CURRENT STATE  ERROR


docker node ps --filter name=web.1 self
ID  NAME  IMAGE  NODE  DESIRED STATE  CURRENT STATE  ERROR
```

New behavior:

```
docker node ps self
ID                         NAME        IMAGE         NODE          DESIRED STATE  CURRENT STATE          ERROR
f427p2tgd3z3zdvai81a8eejr  database.1  nginx:alpine  3df591577797  Running        Running 2 minutes ago
en923wd0l1cvpvezay2kw1pcn  web.1       nginx:alpine  3df591577797  Running        Running 6 minutes ago
e4s6pj7r5236bj9mq8rwyjin2  database.2  nginx:alpine  3df591577797  Running        Running 2 minutes ago
5ioynker4k6d1mj3phsvg6gmx  web.2       nginx:alpine  3df591577797  Running        Running 6 minutes ago
532uyf9h5t6bt7oexdl1dguc5  database.3  nginx:alpine  3df591577797  Running        Running 2 minutes ago
27f0c9bj8xu1xbpca95n7t4xl  web.3       nginx:alpine  3df591577797  Running        Running 6 minutes ago
7w517mezwpf8intwnaewmainq  database.4  nginx:alpine  3df591577797  Running        Running 2 minutes ago
9stqaj4d252stcmzcxypqxfmf  web.4       nginx:alpine  3df591577797  Running        Running 6 minutes ago
dc8skyxufdfvsid5mo7xbu4y3  database.5  nginx:alpine  3df591577797  Running        Running 2 minutes ago
cnah3ap18r8ppmq4ocmuiik24  web.5       nginx:alpine  3df591577797  Running        Running 6 minutes ago

docker node ps --filter name=web. self
ID                         NAME   IMAGE         NODE          DESIRED STATE  CURRENT STATE          ERROR
en923wd0l1cvpvezay2kw1pcn  web.1  nginx:alpine  3df591577797  Running        Running 8 minutes ago
5ioynker4k6d1mj3phsvg6gmx  web.2  nginx:alpine  3df591577797  Running        Running 8 minutes ago
27f0c9bj8xu1xbpca95n7t4xl  web.3  nginx:alpine  3df591577797  Running        Running 8 minutes ago
9stqaj4d252stcmzcxypqxfmf  web.4  nginx:alpine  3df591577797  Running        Running 8 minutes ago
cnah3ap18r8ppmq4ocmuiik24  web.5  nginx:alpine  3df591577797  Running        Running 8 minutes ago

docker node ps --filter name=web.1 self
ID                         NAME   IMAGE         NODE          DESIRED STATE  CURRENT STATE          ERROR
en923wd0l1cvpvezay2kw1pcn  web.1  nginx:alpine  3df591577797  Running        Running 8 minutes ago
```

For `docker service ps`;

Old:

```
docker service ps --filter name=web.1 web

ID  NAME  IMAGE  NODE  DESIRED STATE  CURRENT STATE  ERROR
```

New

```
docker service ps --filter name=web.1 web
ID                         NAME   IMAGE         NODE          DESIRED STATE  CURRENT STATE           ERROR
en923wd0l1cvpvezay2kw1pcn  web.1  nginx:alpine  3df591577797  Running        Running 10 minutes ago
```

**Comment by thaJeztah** [comment]
UX looks good to me

ping @stevvooe PTAL if it looks good on the SwarmKit side :smile:

File: daemon/cluster/executor/container/container.go [hunk 1]
```diff
@@ -72,13 +72,8 @@ func (c *containerConfig) spec() *api.ContainerSpec {
 }
 
 func (c *containerConfig) name() string {
-	if c.task.Annotations.Name != "" {
```
Comment (stevvooe) [inline daemon/cluster/executor/container/container.go]: Don't remove this. We want to be able to name tasks this way in the future. Removing this makes backwards compatibility impossible.

**Comment by stevvooe** [comment]
@thaJeztah Mostly, we just need to make sure these task names are unique. We can either truncate or include a field called `SlotName`. I prefer the former, but the latter is fine.

What I want to avoid is splaying the naming format across the two code bases. The result of this PR should be to concentrate task naming to the orchestrator.

**Comment by thaJeztah** [comment]
@yongtang think you can update the PR with @stevvooe's suggestion?

**Comment by yongtang** [comment]
Thanks @stevvooe @thaJeztah. Let me take a look and update the PR shortly.

**Comment by yongtang** [comment]
@stevvooe @thaJeztah The PR has been updated. Now the Task name is changed to `<service>.<slot>.<taskid>`.

The PR for swarmkit (https://github.com/docker/swarmkit/pull/1193) has been updated as well.

In addition, a pull request for engine-api (https://github.com/docker/engine-api/pull/350) has been created.

Please take a look and let me know if there are any issues.

**Comment by thaJeztah** [comment]
Thanks @yongtang! I'll let @stevvooe have a look first if this is how he had it in mind :+1:

File: daemon/cluster/executor/container/container.go [hunk 2]
```diff
@@ -124,7 +124,7 @@ func (c *containerConfig) labels() map[string]string {
 		system = map[string]string{
 			"task":         "", // mark as cluster task
 			"task.id":      c.task.ID,
-			"task.name":    fmt.Sprintf("%v.%v", c.task.ServiceAnnotations.Name, c.task.Slot),
```
Comment (stevvooe) [inline daemon/cluster/executor/container/container.go]: You need to handle fallback here. if `c.task.Annotations.Name` is not set, fallback to this `fmt.Sprintf` value.

Comment (yongtang) [inline daemon/cluster/executor/container/container.go]: Thanks @stevvooe the pull request has been updated.

**Comment by tiborvass** [comment]
We'll see if we want this for a potential 1.12.2 or a 1.13.0

**Comment by aaronlehmann** [comment]
```
ubuntu@ubuntu:~$ sudo docker service ps --filter name=top.1 top
ID                         NAME                                 IMAGE    NODE    DESIRED STATE  CURRENT STATE                ERROR
9k0jjra8lqipvk7k651hkr9h1  top.1.9k0jjra8lqipvk7k651hkr9h1      busybox  ubuntu  Running        Running about a minute ago   
3ue05jmafu3o03lxdjo4k2fd6   \_ top.1.3ue05jmafu3o03lxdjo4k2fd6  busybox  ubuntu  Shutdown       Complete about a minute ago  
a95stag9gel17qhfk81wcv4nq   \_ top.1.a95stag9gel17qhfk81wcv4nq  busybox  ubuntu  Shutdown       Complete 19 minutes ago   
```

I wonder if it's worth showing a name column (by default), since now the name is quite large, and it's somewhat redundant with information available in other columns. Maybe changing this to a slot column would make sense? (Just an idea - wait until there's more discussion to make a change like this)

Otherwise LGTM.

**Comment by stevvooe** [comment]
> Maybe changing this to a slot column would make sense? 

Or get rid of ID?

**Comment by stevvooe** [comment]
LGTM

**Comment by yongtang** [comment]
Thanks @stevvooe @aaronlehmann. Maybe removing `ID` might be an option, as `NAME` already contains all the information (and unique).

Separately, I have vendored swarmkit in the pull request. Now this pull request still need engine-api update in:
https://github.com/docker/engine-api/pull/350

Please take a look and let me know if there are any issues.

**Comment by yongtang** [comment]
The engine-api has been vendored. The failing test of experimental depends on #25424. Once #25424 is merged in, the tests should pass.

**Comment by vdemeester** [comment]
@yongtang needs a rebase 👼

**Comment by yongtang** [comment]
Thanks @vdemeester. The pull request has been updated with engine-api (https://github.com/docker/engine-api/pull/350) revendored. Now all tests passed.

**Comment by vdemeester** [comment]
@yongtang arf need another rebase :sweat: 
/ping @aaronlehmann

**Comment by vdemeester** [comment]
Code LGTM 🐮

**Comment by yongtang** [comment]
Thanks @vdemeester @aaronlehmann the pull request has been rebased.

**Comment by aaronlehmann** [comment]
What happens to the naming of global service tasks? It looks like they end up with the name `<servicename>.0.<taskid>`, since global services do not use slot numbers. It seems strange for there to be a 0 in the middle. It would make more sense if global tasks were named with `<servicename>.<nodeid>.<taskid>` (but this would produce a long name).

I should have brought this up before the swarmkit PR was merged, since this is really on the swarmkit side.

Also, I really think we should change the `docker service ps` UI now that this PR is adding the task ID to task names. Showing both the name and ID would look very redundant and require a large terminal window. I think we should remove either the ID or name from the default output - probably remove the ID and keep the name.

**Comment by yongtang** [comment]
Thanks @aaronlehmann. Let me take a look and create pull request based on the discussion.

**Comment by aluzzardi** [comment]
> Thanks @stevvooe @aaronlehmann. Maybe removing ID might be an option, as NAME already contains all the information (and unique).

Yeah, that was the entire goal.  The `NAME` column is syntactic sugar in the CLI and does not map the data 1:1 on purpose.

Displaying the `ID` is redundant (there's already an `ID` column), so we could either have two columns (`SERVICE NAME` and `SLOT` - which we used to have), but a `NAME` column that combines `<SERVICE NAME>.<SLOT>` just looks better.

**Comment by vdemeester** [comment]
@yongtang needs a rebase :sweat:

**Comment by yongtang** [comment]
Thanks @vdemeester the pull request has been rebased.

**Comment by yongtang** [comment]
The pull request has also been updated and `ID` has been removed from the output. Now the output is:
```
NAME                                           IMAGE    NODE          DESIRED STATE  CURRENT STATE                     ERROR
redis-cluster-md5.1.4zzvdu001pwy53sjy37b0tpmh  busybox  a5da3f1a8eaa  Running        Allocated less than a second ago  
redis-cluster-md5.2.c9f64oy0c862c5kpo5rvz417t  busybox  a5da3f1a8eaa  Running        Allocated less than a second ago  
redis-cluster-md5.3.52rnsfi4gcv6vc868pz7xc1e1  busybox  a5da3f1a8eaa  Running        Allocated less than a second ago  
```

This pull request still depends on https://github.com/docker/swarmkit/pull/1476

**Comment by nishanttotla** [comment]
@yongtang @vdemeester do we need more review to merge this PR?

**Comment by yongtang** [comment]
@nishanttotla This PR is related to a swarmkit PR https://github.com/docker/swarmkit/pull/1476.
This PR assumes task name to be defined as `<svc name>.<slot>.<task id>`.

In https://github.com/docker/swarmkit/pull/1476 I think @aluzzardi suggest to use DNS notation for task name `<task id>.<slot>.<svc name>`.

I will update the PR in https://github.com/docker/swarmkit/pull/1476 and hopefully we could get both merged soon.

**Comment by nishanttotla** [comment]
Thanks @yongtang. I will review the SwarmKit PR, and hope we can merge that soon, to unblock this one.

**Comment by nishanttotla** [comment]
@yongtang sorry this will need to be rebased.

**Comment by yongtang** [comment]
Thanks @nishanttotla. The pull request has been rebased and swarmkit has been vendored in. Please take a look and let me know if there are any issues.

**Comment by thaJeztah** [comment]
FWIW, this SwarmKit vendor brings in a lot of changes that will be used for a feature that @ehazlett is working on; I briefly discussed if we should try and get those changes in as part of _that_ PR, but that would unnecessarily delay other PRs, so should not be a blocking issue.

**Comment by aaronlehmann** [comment]
LGTM

**Review by vdemeester — APPROVED** [review]
LGTM 🐸

**Comment by vdemeester** [comment]
@thaJeztah needs some doc updates ?

**Comment by thaJeztah** [comment]
@vdemeester I'll have to check if we mention that only "full" matches were supported

**Review by thaJeztah — APPROVED** [review]
I think docs don't need updating for this, so LGTM

**Comment by thaJeztah** [comment]
oh, hold on, we dropped the `ID` column now. Checking again, (sorry)

**Review by thaJeztah — CHANGES_REQUESTED** [review]
@yongtang looks like the output examples in 
- https://github.com/docker/docker/blob/master/docs/reference/commandline/node_ps.md,
- https://github.com/docker/docker/blob/master/docs/reference/commandline/service_ps.md, 
- https://github.com/docker/docker/blob/master/docs/swarm/networking.md#attach-a-service-to-an-overlay-network
- https://github.com/docker/docker/blob/master/docs/swarm/swarm-tutorial/drain-node.md#drain-a-node-on-the-swarm
- https://github.com/docker/docker/blob/master/docs/swarm/swarm-tutorial/inspect-service.md
- https://github.com/docker/docker/blob/master/docs/swarm/swarm-tutorial/rolling-update.md
- https://github.com/docker/docker/blob/master/docs/swarm/swarm-tutorial/scale-service.md

Need to be updated due to the output changing (and `ID` being dropped). Sorry, it's quite a long list :cry:

**Comment by yongtang** [comment]
Thanks @thaJeztah. I have updated the PR for the docs changes. There are some additional change other than removing `ID` as some of the sample output was using very old format. Those have been fixed as well.

Please take a look and let me know if there are additional issues.

**Review by thaJeztah — APPROVED** [review]
perfect, LGTM

**Comment by vdemeester** [comment]
docs LGTM 🐸

**Comment by thaJeztah** [comment]
Experimental is a known flaky test; https://github.com/docker/docker/issues/26588

WindowsRS1 looks unrelated, but not sure we have an issue tracking that already;

```
04:18:03 ----------------------------------------------------------------------
04:18:03 FAIL: docker_cli_run_test.go:1855: DockerSuite.TestRunInteractiveWithRestartPolicy
04:18:03 
04:18:03 docker_cli_run_test.go:1868:
04:18:03     c.Assert(result, icmd.Matches, icmd.Expected{ExitCode: 11})
04:18:03 ... result *cmd.Result = &cmd.Result{Cmd:(*exec.Cmd)(0xc04221f1e0), ExitCode:0, Error:error(nil), Timeout:true, outBuffer:(*cmd.lockedBuffer)(0xc04219aab0), errBuffer:(*cmd.lockedBuffer)(0xc04219ab40)} ("\nCommand: d:\\CI\\CI-47ac185\\binary\\docker.exe run -i --name test-inter-restart --restart=always busybox sh\nExitCode: 0 (timeout), Error: %!s(<nil>)\nStdout: \nStderr: \n")
04:18:03 ... expected cmd.Expected = cmd.Expected{ExitCode:11, Timeout:false, Error:"", Out:"", Err:""}
04:18:03 ... 
04:18:03 Command: d:\CI\CI-47ac185\binary\docker.exe run -i --name test-inter-restart --restart=always busybox sh
04:18:03 ExitCode: 0 (timeout), Error: %!s(<nil>)
04:18:03 Stdout: 
04:18:03 Stderr: 
04:18:03 
04:18:03 Failures:
04:18:03 ExitCode was 0 expected 11
04:18:03 Expected command to finish, but it hit the timeout
```

**Comment by thaJeztah** [comment]
Thanks @yongtang! It took a while, but we got there 🎉

**Pull request merged**